### ISSUE TITLE
tm.connections.destroySegment and tm.connections.destroySynapse do not seem to work

### ISSUE THREAD

**Issue by mohdsherif** [htm-community/htm.core#911]
Hi,

After initializing a TM using default parameters and training it, I get 900 synapses on 300 segments. 
When I try to destroy a segment or a synapse, I still get the same number of synapses/segments. Also, when I check the permanence, the permanence of the synapses I destroyed have not changed. Am I missing a step I should be doing after calling tm.connections.destroySegment() or tm.connections.destroySynapse()? Here is a MWE:

```python
from htm.algorithms import TemporalMemory as TM
from htm.bindings.sdr import SDR

columnDimensions = [2048] 
cellsPerColumn = 32
activationThreshold = 15
initialPermanence = 0.21
connectedPermanence = 0.3
minThreshold = 15
maxNewSynapseCount = 40
permanenceIncrement = 0.1
permanenceDecrement = 0.1
predictedSegmentDecrement = 0.01
seed = 42
maxSegmentsPerCell = 255
maxSynapsesPerSegment = 255
checkInputs = False # Whether to check that the activeColumns are sorted without duplicates.
externalPredictiveInputs = 0
# anomalyMode: htm.bindings.algorithms.ANMode = ANMode.RAW) -> None

tm = TM(columnDimensions, cellsPerColumn, activationThreshold, initialPermanence, connectedPermanence, minThreshold, maxNewSynapseCount, permanenceIncrement, permanenceDecrement, predictedSegmentDecrement, seed, maxSegmentsPerCell, maxSynapsesPerSegment, checkInputs, externalPredictiveInputs)

lbits = [[1, 45, 100], [300, 340, 900], [3, 2, 100]]
lSDR = []
for bits in lbits:
    mySDR = SDR(tm.getColumnDimensions()[0])
    mySDR.sparse = list(bits)
    lSDR.append(mySDR)

for i in range(50):
    for mySDR in lSDR:
        tm.compute(mySDR, learn=True)
        tm.activateDendrites(True)
    tm.reset()

# number of synapses 
tm.connections.numSynapses() # 900
# destroy all synapses
for mysyn in range(tm.connections.numSynapses()): tm.connections.destroySynapse(mysyn)
# number of synapses is still 900
tm.connections.numSynapses()

# the same with segments
tm.connections.numSegments() # 300
# destroy all segments
for myseg in range(tm.connections.numSegments()): tm.connections.destroySegment(myseg)
tm.connections.numSegments() # still 300
```

Thank you for your help :-)

**Comment by ctrl-z-9000-times** [comment]
Hi,

I looked into this and I am able to reproduce your results.
There could be a bug in `connections.numSynapses()`?

> the permanence of the synapses I destroyed have not changed

The Connections class should not overwrite any data when it deletes a synapse, instead it unlinks the synapse in the fastest way it can, which leaves the old data in memory.

The connections class is not intended to be directly used. Instead it exists to help the SpatialPooler & TemporalMemory algorithms, and my advice is to avoid the Connections class unless you are developing an algorithm like the SP or TM which use lots of synapses.

**Comment by mohdsherif** [comment]
Thanks a lot for looking into this.

I see. My concern is that also the network does not "forget". I was able to predict learned patterns even after destroying all synapses. Am I missing something here?

Regarding your point that the Connections class is not intended for direct use, is there a recommended way to manipulate the synapses and segments directly other than the connections class?

Thanks again.

**Comment by ctrl-z-9000-times** [comment]
> I was able to predict learned patterns even after destroying all synapses. 

There is probably a bug in the Connections/TemporalMemory implementation.

> is there a recommended way to manipulate the synapses and segments directly other than the connections class?

Ideally, you'd use the TemporalMemory without accessing the underlying Connections object. The temporal memory does have a number of features which make it quite versatile as a general purpose model for synapses. In specific, you can:
- Set the number of cells per mini-column to 1,
- Make predictions based off of external inputs,
- Compute the predicted cells before deciding which cells activate.

**Comment by mohdsherif** [comment]
Thanks again for taking the time. I appreciate the input.

I can start looking into the bug - any suggestions for which files to start looking at?

**Comment by ctrl-z-9000-times** [comment]
> I can start looking into the bug

I think that the first thing to do would be to make a test case for it so that you'll know when it's fixed (and so that it stays fixed).

It would be most convincing if you demonstrated that the TM fails under seemingly normal usage (without manually altering the underlying connections). Maybe make a very small TM and use up every synapse, and then test that it can still learn new things by removing the existing synapses?

**Comment by mohdsherif** [comment]
Sounds good - will start looking into this - thanks for the pointers.

**Comment by ctrl-z-9000-times** [comment]
Hi, I think finally figured out why this does not work: the connections object is read only.

It is defined as a private variable in C++, with a read-only public accessor.
Python likewise can read, but not modify, to the connections object.

I'm pretty sure that when you type `tm.connections.foobar()` that under the hood pybind11 is creating a full copy of the connections object for you to play with, and that it is not giving you a reference to the TM's instance of Connections.

The htm.core library is just not capable of doing the things in your example.
In summary:
- You can make and modify your own instance of Connections,
- You can inspect the TM's connections,
- You can *not* modify the TM's connections.

I'm going to close this issue. If you have more questions feel free to open another issue.
